JS Basics/Tetris/index.js

```diff
@@ -0,0 +1,569 @@
+      return moveRight ? pointX < maxX : pointX > minX;
+    }
+  });
+}
+
+function moveShapeDown() {
+  if (canMoveDown() && !checkForCollision()) {
+    currentShape.points.forEach((point) => {
+      point.coordinates.y += 1;
+      setPosition(point.domElement, point.coordinates);
+    });
+  }
+}
+
+function canMoveDown() {
+  return currentShape.points.every((point) => {
+    const pointY = point.coordinates.y;
+    return pointY < maxY;
+  });
+}
+
+function rotateShape() {
+  getShapeRotation();
+  if (!canRotateShape()) {
+    const rotationOffset = getOffset();
+    currentShape.points.forEach((point) => {
+      point.coordinates.x += rotationOffset;
+    });
+  }
+}
+
+function canRotateShape() {
+  return currentShape.points.every((point) => {
+    let pointXMax = point.coordinates.x;
+    let pointXMin = point.coordinates.x;
+
+    return pointXMax <= maxX && pointXMin >= minX;
+  });
+}
+
+function getOffset() {
+  let offset = 0;
+
+  currentShape.points.some((point) => {
+    const pointX = point.coordinates.x;
+
+    if (pointX === maxX) {
+      if (currentShape.type === TETRIS_SHAPE.I) {
+        offset = -2;
+      } else {
+        offset = -1;
+      }
+    } else if (pointX === minX) {
+      offset = 1;
+    }
+  });
+
+  return offset;
+}
+
+function getShapeRotation() {
+  switch (currentShape.type) {
+    case TETRIS_SHAPE.I:
+      rotateIShape();
+      break;
+    case TETRIS_SHAPE.J:
+      rotateJShape();
+      break;
+    case TETRIS_SHAPE.L:
+      rotateLShape();
+      break;
+    case TETRIS_SHAPE.S:
+      rotateSShape();
+      break;
+    case TETRIS_SHAPE.T:
+      rotateTShape();
+      break;
+    case TETRIS_SHAPE.Z:
+      rotateZShape();
+      break;
+  }
+}
+
+function rotateIShape() {
+  if (currentShape.hasBeenRotatedOnce) {
+    currentShape.points[0].coordinates.x += 1;
+    currentShape.points[0].coordinates.y -= 1;
+
+    currentShape.points[2].coordinates.x -= 1;
+    currentShape.points[2].coordinates.y += 1;
+
+    currentShape.points[3].coordinates.x -= 2;
+    currentShape.points[3].coordinates.y += 2;
+
+    currentShape.hasBeenRotatedOnce = false;
+  } else {
+    currentShape.points[0].coordinates.x -= 1;
+    currentShape.points[0].coordinates.y += 1;
+
+    currentShape.points[2].coordinates.x += 1;
+    currentShape.points[2].coordinates.y -= 1;
+
+    currentShape.points[3].coordinates.x += 2;
+    currentShape.points[3].coordinates.y -= 2;
+
+    currentShape.hasBeenRotatedOnce = true;
+    return;
+  }
+}
+
+function rotateJShape() {
+  if (currentShape.hasBeenRotatedOnce) {
+    currentShape.points[0].coordinates.x -= 1;
+    currentShape.points[0].coordinates.y += 1;
+
+    currentShape.points[2].coordinates.x += 1;
+    currentShape.points[2].coordinates.y -= 1;
+
+    currentShape.points[3].coordinates.x -= 2;
+
+    currentShape.hasBeenRotatedOnce = false;
+    currentShape.hasBeenRotatedTwice = true;
+    currentShape.hasBeenRotatedThrice = false;
```

Comment:
Instead of having three bolean variables, would not it be easier to have one integer i.e. numberOfRotations: 1 | 2 | 3 ?